src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,672 @@
+                                mr);
+  // column to store the status `ParseResult`
+  auto result_valid_col = cudf::make_fixed_width_column(
+    cudf::data_type{cudf::type_id::UINT8}, input.size(), cudf::mask_state::UNALLOCATED, stream, mr);
+
+  if (transitions == nullptr || tz_indices == nullptr) {
+    thrust::transform(
+      rmm::exec_policy(stream),
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(input.size()),
+      thrust::make_zip_iterator(
+        thrust::make_tuple(result_col->mutable_view().begin<cudf::timestamp_us>(),
+                           result_valid_col->mutable_view().begin<uint8_t>())),
+      parse_timestamp_string_fn<false>{*d_strings, default_tz_index, allow_tz_in_date_str});
+  } else {
+    auto const ft_cdv_ptr    = column_device_view::create(*transitions, stream);
+    auto const d_transitions = lists_column_device_view{*ft_cdv_ptr};
+    auto d_tz_indices        = cudf::column_device_view::create(*tz_indices, stream);
+
+    thrust::transform(rmm::exec_policy(stream),
+                      thrust::make_counting_iterator(0),
+                      thrust::make_counting_iterator(input.size()),
+                      thrust::make_zip_iterator(
+                        thrust::make_tuple(result_col->mutable_view().begin<cudf::timestamp_us>(),
+                                           result_valid_col->mutable_view().begin<uint8_t>())),
+                      parse_timestamp_string_fn<true>{
+                        *d_strings, default_tz_index, true, d_transitions, *d_tz_indices});
+  }
+
+  auto valid_view = result_valid_col->mutable_view();
+
+  // throw cuDF exception if there exists any unsupported formats
+  auto exception_exists =
+    thrust::any_of(rmm::exec_policy(stream),
+                   valid_view.begin<uint8_t>(),
+                   valid_view.end<uint8_t>(),
+                   [] __device__(uint8_t e) { return e == ParseResult::UNSUPPORTED; });
+  if (exception_exists) { CUDF_FAIL("There exists unsupported timestamp schema!"); }
+
+  // build the updated null mask and compute the null count
+  auto [valid_bitmask, valid_null_count] = cudf::detail::valid_if(
+    valid_view.begin<uint8_t>(),
+    valid_view.end<uint8_t>(),
+    [] __device__(uint8_t e) { return e == 0; },
+    stream,
+    mr);
+
+  // `output null count > input null count` indicates that there are new null
+  // values generated during the `to_timestamp` transaction to replace invalid
+  // inputs.
+  if (ansi_mode && input.null_count() < valid_null_count) { return nullptr; }
+
+  result_col->set_null_mask(valid_bitmask, valid_null_count, stream);
+  return result_col;
+}
+
+}  // namespace
+
+namespace spark_rapids_jni {
+
+/**
+ * Parse string column with time zone to timestamp column.
+ * If a string does not have time zone in it, use the default time zone.
+ * Returns nullptr if ANSI mode is true and strings have any invalid value, returns non-null
+ * timestamp column otherwise.
+ *
```

Comment:
What happens if you have invalid data and do not set ANSI mode? Maybe something like 
```
 * If strings have invalid data, return nullptr is ANSI mode is true or empty string if ANSI mode is false.

Reply:
```
    auto ret_cv         = spark_rapids_jni::string_to_timestamp_with_tz(
  ......
    if (ret_cv) { return cudf::jni::release_as_jlong(ret_cv); }
  ......
  JNI_THROW_NEW(env, "java/lang/IllegalArgumentException", "Parse failed on Ansi mode", 0);
```

```
 * Returns nullptr if ANSI mode is true and strings have invalid data,
 * otherwise, returns non-null timestamp column(the invalid date will be empty in this column)
```